### PR TITLE
update contribution guidelines

### PR DESCRIPTION
**Pull request by epwalsh**
Given all of the new checks we now run on pull requests, I think we really need to strive to ensure the contribution process remains friendly and straight-forward for collaborators.

Let me know if you think I should add or change anything here.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -56,14 +36,129 @@ When creating your enhancement request, please:
...
+        git remote add upstream https://github.com/allenai/allennlp.git
+    
+    Now if you do `git remote -v` again, you'll see
+    
+        origin https://github.com/USERNAME/allennlp.git (fetch)
+        origin https://github.com/USERNAME/allennlp.git (push)
+        upstream https://github.com/allenai/allennlp.git (fetch)
+        upstream https://github.com/allenai/allennlp.git (push)
+
+    Finally, you'll need to create a Python 3.6 or 3.7 virtual environment suitable for working on AllenNLP. There a number of tools that provide
```
Comment (matt-gardner) [inline CONTRIBUTING.md:77]: Incomplete sentence.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -56,14 +36,129 @@ When creating your enhancement request, please:
...
 
-We are always looking for new models to add to our collection.  If you have trained a model and would like to include it in
-AllenNLP, please create [a pull request](https://github.com/allenai/allennlp/pulls) that includes:
+    Commiting directly to the master branch of your fork is not recommended. It will be easier to keep your fork clean if you work on a seperate branch for each contribution you intend to make.
+    
+    You can create a new branch with
+    
+        # replace BRANCH with whatever name you want to give it
+        git checkout -b BRANCH
+        git push --set-upstream origin BRANCH
```
Comment (matt-gardner) [inline CONTRIBUTING.md:103]: Or just `git push -u origin BRANCH`

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -56,14 +36,129 @@ When creating your enhancement request, please:
...
+
+    Our CI also uses [`flake8`](https://github.com/allenai/allennlp/tree/master/tests) to lint the code base and [`mypy`](http://mypy-lang.org/) for type-checking. You should run both of these next with
+
+        make lint
+
+    and
+
+        make typecheck
+
+    We also strive to maintain high test coverage, so most contributions should include additions to [the unit tests](https://github.com/allenai/allennlp/tree/master/tests). These tests are ran with [`pytest`](https://docs.pytest.org/en/latest/), which you can use to locally run any test modules that you've added or changed.
```
Comment (matt-gardner) [inline CONTRIBUTING.md:125]:
```suggestion
    We also strive to maintain high test coverage, so most contributions should include additions to [the unit tests](https://github.com/allenai/allennlp/tree/master/tests). These tests are run with [`pytest`](https://docs.pytest.org/en/latest/), which you can use to locally run any test modules that you've added or changed.
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -56,14 +36,129 @@ When creating your enhancement request, please:
...
+
+    For example, if you've fixed a bug in `allennlp/nn/util.py`, you can run the tests specific to that module with
+    
+        pytest -v tests/nn/util_test.py
+    
+    Our CI will automatically check that test coverage stays above a certain threshold (around 90%). To check the coverage locally in this example, you could run
+    
+        pytest -v --cov allennlp.nn.util tests/nn/util_test.py
+
+    You should also make sure that the API documentation can build without errors. For that, just run
```
Comment (matt-gardner) [inline CONTRIBUTING.md:135]: Maybe some note here that failures here are pretty rare, and you probably won't have to run this.  But if you need to, you can run it locally with this.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -56,14 +36,129 @@ When creating your enhancement request, please:
...
+
+    We look forward to reviewing your PR!
+
+    </details>
+
+### New models
+
+**Do you have a new state-of-the-art model?**
+
+We are always looking for new models to add to our collection.  If you have trained a model and would like to include it with
```
Comment (matt-gardner) [inline CONTRIBUTING.md:156]: I'm not actually sure this should be our recommendation.  Having other people's models live in third party repositories is a solution we should encourage.

File: CONTRIBUTING.md (hunk 1 above)
Comment (epwalsh) [reply CONTRIBUTING.md:77]: Oops! Good catch

File: CONTRIBUTING.md (hunk 5 above)
Comment (epwalsh) [reply CONTRIBUTING.md:156]: Should we just remove this whole section then?

Comment (matt-gardner) [reply CONTRIBUTING.md:156]: I think this is a question that we should answer.  I'm just not sure what the answer should be, and I would lean towards the answer being, "we add your model to a list of third-party repos that we display prominently somewhere".  In certain cases, we can add them to our models repo, or offer to have them in our demo, or something, but that's not necessary.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -56,14 +36,138 @@ When creating your enhancement request, please:
...
+    After all of the above checks have passed, you can now open [a new GitHub pull request](https://github.com/allenai/allennlp/pulls).
+    Make sure you have a clear description of the problem and the solution, and include a link to relevant issues.
+
+    We look forward to reviewing your PR!
+
+    </details>
+
+### New models
+
+**Do you have a new state-of-the-art model?**
```
Comment (epwalsh) [inline CONTRIBUTING.md:162]: @matt-gardner you may want to take another look at this now

Comment (matt-gardner) [reply CONTRIBUTING.md:162]: Sure, seems ok.  I think eventually we'll want a better story here, but we don't currently have one, so this is good enough.

File: CONTRIBUTING.md (hunk 5 above)
Comment (dirkgr) [reply CONTRIBUTING.md]: One of our longer-term goals post-1.0 (and post-performance) is to have a better story for third-party models. For now, if someone has a cool one, we should at least keep track of it somewhere, maybe in an issue, so we can take it into account when we have time to tackle this.

Comment (epwalsh) [reply CONTRIBUTING.md]: How about we create an issue and link to it from `CONTRIBUTING.md` and maybe also the main `README.md`?

**Pull request merged**